### PR TITLE
feat: implementing #65 :

### PR DESCRIPTION
**Pull request by dhruvmalik007**
Author: Dhruv (@dhruvmalik007 ).

- Writing initial version of the kernel inception distance metric (with comparison to the torch-metrics).
- Writing the test cases to cover edge case for classification.

### CODE REVIEW

**Comment by jshin1394** [comment]
Hi Dhruv, thank you so much for your contribution to the Metrax repo! The first iteration of your PR looks very promising! :)

A few things:
1. Could you please sign the Google Contributor License Agreement(CLA) as suggested by the google-cla bot?
2. Here is an example PR which adds a new metric to Metrax(https://github.com/google/metrax/pull/79/files). It also adds the metric to `metrax_test` in order to test its jittability and adds the metrax to metrax/nnx in order to support NNX users.
3. clu_metric.Metric interface has a list of functions that must be implemented and merge is one of them(https://github.com/google/metrax/pull/79/files).

Thank you so much! :)

File: src/metrax/image_metrics.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+KID_DEFAULT_SUBSETS = 100
+KID_DEFAULT_SUBSET_SIZE = 1000
+KID_DEFAULT_DEGREE = 3
+KID_DEFAULT_GAMMA = None
+KID_DEFAULT_COEF = 1.0
+
+
+@flax.struct.dataclass
+class KernelInceptionDistanceMetric(clu_metrics.Metric):
```
Comment (jshin1394) [inline src/metrax/image_metrics.py:32]: what are your thoughts on having this inherit base.Average from the metrax package? It will keep track of the sum of total KID and the sum of total count and compute will yield the average KID of the data.

File: src/metrax/image_metrics.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+        kt_xx_sum = jnp.sum(k_11, axis=-1) - diag_x
+        kt_yy_sum = jnp.sum(k_22, axis=-1) - diag_y
+        k_xy_sum = jnp.sum(k_12, axis=0)
+
+        value = (jnp.sum(kt_xx_sum) + jnp.sum(kt_yy_sum)) / (m * (m - 1))
+        value -= 2 * jnp.sum(k_xy_sum) / (m**2)
+        return value
+
+    def polynomial_kernel(self, x: jax.Array, y: jax.Array) -> jax.Array:
```
[the remark this answers is not in the document]
Comment (dhruvmalik007) [reply src/metrax/image_metrics.py:134]: Sounds good, will declare it outside the class and then reference it in private function and future implementations. Normally we can define utils.py in order to aggregate all of these composable functions.

File: src/metrax/image_metrics.py (hunk 1 above)
Comment (dhruvmalik007) [reply src/metrax/image_metrics.py]: indeed I have done that and during my prompting , I realised to implement it in the following function: 

```python
    @classmethod
    def from_model_output(
        cls,
        real_features: jax.Array,
        fake_features: jax.Array,
        subsets: int = KID_DEFAULT_SUBSETS,
        subset_size: int = KID_DEFAULT_SUBSET_SIZE,
        degree: int = KID_DEFAULT_DEGREE,
        gamma: float = KID_DEFAULT_GAMMA,
        coef: float = KID_DEFAULT_COEF,
    ):
        # checks for the valid inputs
        if subsets <= 0 or subset_size <= 0 or degree <= 0 or (gamma is not None and gamma <= 0) or coef <= 0:
            raise ValueError("All parameters must be positive and non-zero.")
        # Compute KID for this batch
        if real_features.shape[0] < subset_size or fake_features.shape[0] < subset_size:
            raise ValueError("Subset size must be smaller than the number of samples.")
        master_key = random.PRNGKey(42)
        kid_scores = []
        for i in range(subsets):
            key_real, key_fake = random.split(random.fold_in(master_key, i))
            real_indices = random.choice(key_real, real_features.shape[0], (subset_size,), replace=False)
            fake_indices = random.choice(key_fake, fake_features.shape[0], (subset_size,), replace=False)
            f_real_subset = real_features[real_indices]
            f_fake_subset = fake_features[fake_indices]
            kid = cls.__compute_mmd_static(f_real_subset, f_fake_subset, degree, gamma, coef)
            kid_scores.append(kid)
        kid_mean = jnp.mean(jnp.array(kid_scores))
        # Accumulate sum and count for averaging
        return cls(
            total=kid_mean,
            count=1.0,
            subsets=subsets,
            subset_size=subset_size,
            degree=degree,
            gamma=gamma,
            coef=coef,
        )
```

**Comment by dhruvmalik007** [comment]
> Thank you so much Dhruv!

your welcome and appreciated your help for resolving issues.

File: src/metrax/image_metrics.py [hunk 3]
```diff
@@ -53,6 +62,181 @@ def _gaussian_kernel1d(sigma, radius):
...
+            count=0.0,
+            subsets=KID_DEFAULT_SUBSETS,
+            subset_size=KID_DEFAULT_SUBSET_SIZE,
+            degree=KID_DEFAULT_DEGREE,
+            gamma=KID_DEFAULT_GAMMA,
+            coef=KID_DEFAULT_COEF,
+        )
+
+    @staticmethod
+    def __compute_mmd_static(f_real: jax.Array, f_fake: jax.Array, degree: int, gamma: float, coef: float) -> float:
```
Comment (jshin1394) [inline src/metrax/image_metrics.py:195]: I think single underscore is fine here?

**Review by jshin1394 — COMMENTED** [review]
Thank you :)

File: src/metrax/image_metrics_test.py [hunk 4]
```diff
@@ -66,115 +73,359 @@
...
+
+        kid_mean_torch, kid_std_torch, kid_mean_metrax, kid_std_metrax = self.compute_torchmetrics_kid(
+            imgs_real, imgs_fake, subsets=subsets, subset_size=subset_size
+        )
+        kid_metric = KID.from_model_output(
+            jnp.array(real_features), jnp.array(fake_features),
+            subsets=subsets, subset_size=subset_size
+        )
+        # Accept numpy scalar or float
+        ## return float(kid_mean.cpu().numpy()), float(kid_std.cpu().numpy()), metrax_mean, metrax_std
```
Comment (jshin1394) [inline src/metrax/image_metrics_test.py:154]: nit: remove

File: src/metrax/image_metrics_test.py [hunk 5]
```diff
@@ -66,115 +73,359 @@
...
+        subset_size: int = 16
+        imgs_real = random_images(0, n)
+        imgs_fake = random_images(1, n)
+        real_features = np.random.randn(n, 2048).astype(np.float32)
+        fake_features = np.random.randn(n, 2048).astype(np.float32)
+
+        kid_mean_torch, kid_std_torch, kid_mean_metrax, kid_std_metrax = self.compute_torchmetrics_kid(
+            imgs_real, imgs_fake, subsets=subsets, subset_size=subset_size
+        )
+        kid_metric = KID.from_model_output(
```
Comment (jshin1394) [inline src/metrax/image_metrics_test.py:149]: do we actully compare the value against torchmetrics?

**Review by jshin1394 — COMMENTED** [review]
Thank you so much :)

File: src/metrax/image_metrics_test.py (hunk 5 above)
Comment (dhruvmalik007) [reply src/metrax/image_metrics_test.py:270]: Thanks for noting,  so indeed the `compute_torchmetrics` static function didnt had correct implementation. I have refactored that  in the current version.

File: src/metrax/nnx/__init__.py [hunk 6]
```diff
@@ -62,4 +63,5 @@
     "RougeN",
     "SSIM",
     "WER",
+    "KID"
```
Comment (jshin1394) [inline src/metrax/nnx/__init__.py:66]: same.

File: src/metrax/image_metrics.py [hunk 7]
```diff
@@ -53,6 +60,119 @@ def _gaussian_kernel1d(sigma, radius):
...
+            count=1.0,
+            subsets=subsets,
+            subset_size=subset_size,
+            degree=degree,
+            gamma=gamma,
+            coef=coef,
+        )
+
+
+    @staticmethod
```
Comment (jshin1394) [inline src/metrax/image_metrics.py:158]: let's move this above from_model_output

**Review by jshin1394 — COMMENTED** [review]
Thank you so much :)

File: src/metrax/image_metrics.py [hunk 8]
```diff
@@ -53,6 +60,119 @@ def _gaussian_kernel1d(sigma, radius):
...
+    degree: int = KID_DEFAULT_DEGREE
+    gamma: float = KID_DEFAULT_GAMMA
+    coef: float = KID_DEFAULT_COEF
+
+    @classmethod
+    def from_model_output(
+        cls,
+        real_features: jax.Array,
+        fake_features: jax.Array,
+        subsets: int = KID_DEFAULT_SUBSETS,
```
[the remark this answers is not in the document]
Comment (dhruvmalik007) [reply src/metrax/image_metrics.py:120]: Sounds good. It was just in case if some `model_eval.py` evaluation script might import this parameter .

**Comment by dhruvmalik007** [comment]
hi @jshin1394 , apologies of not notifying on this branch, let me know if are there any remaining issues to be resolved .

**Comment by dhruvmalik007** [comment]
> > hi @jshin1394 , apologies of not notifying on this branch, let me know if are there any remaining issues to be resolved .
> 
> Hi Dhruv, are we assuming that users will call metrax.KID with extracted features only? It seems like torchmetrics.KID takes in as input the real image and the generate image but metrax.KID takes in real_features and generated_features.

Yes you're right .  initially I have taken the I{Real} and I{Fake} as in the reference implementation [here](https://github.com/Lightning-AI/torchmetrics/blob/74bdb26b35515c36334f59d654f0f4d8a12d5b59/src/torchmetrics/image/kid.py#L177-L308) does indeed take the reference parameters.

**Comment by jshin1394** [comment]
> > > hi @jshin1394 , apologies of not notifying on this branch, let me know if are there any remaining issues to be resolved .
> > 
> > 
> > Hi Dhruv, are we assuming that users will call metrax.KID with extracted features only? It seems like torchmetrics.KID takes in as input the real image and the generate image but metrax.KID takes in real_features and generated_features.
> 
> Yes you're right . initially I have taken the I{Real} and I{Fake} as in the reference implementation [here](https://github.com/Lightning-AI/torchmetrics/blob/74bdb26b35515c36334f59d654f0f4d8a12d5b59/src/torchmetrics/image/kid.py#L177-L308) does indeed take the generated features . I have done the changes accordingly .

I think the original implementation takes in images as input and calls this model(https://github.com/Lightning-AI/torchmetrics/blob/74bdb26b35515c36334f59d654f0f4d8a12d5b59/src/torchmetrics/image/kid.py#L211) to get the features. If the implementation does indeed require pretrained model, I will go ahead and come up with a design for this.
